cmd/clusterctl/pkg/client/config_test.go

```diff
@@ -162,3 +163,302 @@ func Test_clusterctlClient_GetProviderComponents(t *testing.T) {
 		})
 	}
 }
+
+func Test_clusterctlClient_templateOptionsToVariables(t *testing.T) {
+	type args struct {
+		options GetClusterTemplateOptions
+	}
+	tests := []struct {
+		name     string
+		args     args
+		wantVars map[string]string
+		wantErr  bool
+	}{
+		{
+			name: "pass (using KubernetesVersion from template options)",
+			args: args{
+				options: GetClusterTemplateOptions{
+					ClusterName:              "foo",
+					TargetNamespace:          "bar",
+					KubernetesVersion:        "v1.2.3",
+					ControlPlaneMachineCount: 1,
+					WorkerMachineCount:       2,
+				},
+			},
+			wantVars: map[string]string{
+				"CLUSTER_NAME":               "foo",
+				"NAMESPACE":                  "bar",
+				"KUBERNETES_VERSION":         "v1.2.3",
+				"CONTROLPLANE_MACHINE_COUNT": "1",
+				"WORKER_MACHINE_COUNT":       "2",
+			},
+			wantErr: false,
+		},
+		{
+			name: "pass (using KubernetesVersion from env variables)",
+			args: args{
+				options: GetClusterTemplateOptions{
+					ClusterName:              "foo",
+					TargetNamespace:          "bar",
+					KubernetesVersion:        "", // empty means to use value from env variables/config file
+					ControlPlaneMachineCount: 1,
+					WorkerMachineCount:       2,
+				},
+			},
+			wantVars: map[string]string{
+				"CLUSTER_NAME":               "foo",
+				"NAMESPACE":                  "bar",
+				"KUBERNETES_VERSION":         "v3.4.5",
+				"CONTROLPLANE_MACHINE_COUNT": "1",
+				"WORKER_MACHINE_COUNT":       "2",
+			},
+			wantErr: false,
+		},
+		{
+			name: "fails for invalid cluster Name",
+			args: args{
+				options: GetClusterTemplateOptions{
+					ClusterName: "A!££%",
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "fails for invalid namespace Name",
+			args: args{
+				options: GetClusterTemplateOptions{
+					ClusterName:     "foo",
+					TargetNamespace: "A!££%",
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "fails for invalid version",
+			args: args{
+				options: GetClusterTemplateOptions{
+					ClusterName:       "foo",
+					TargetNamespace:   "bar",
+					KubernetesVersion: "A!££%",
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "fails for invalid control plane machine count",
+			args: args{
+				options: GetClusterTemplateOptions{
+					ClusterName:              "foo",
+					TargetNamespace:          "bar",
+					KubernetesVersion:        "v1.2.3",
```

Comment:
can `ClusterName`, `TargetNemespace` or `KubernetesVersion` be empty, is that valid? Should we test whatever the desired behavior for that?

Reply:
templateOptionsToVariables handles validation and the defaulting rules when empty is passed. Nevertheless, I'm planning o do a crash test on the UX as soon as we complete all the functionalities